pkg/fileutils/filelock.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fileutils
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"syscall"
+)
+
+// FileLock define a file lock implemented by syscall.Flock
+type FileLock struct {
+	fileName string
+	fd       *os.File
+	mutex    sync.Mutex
+}
+
+// NewFileLock create a FileLock instance
+func NewFileLock(name string) *FileLock {
+	return &FileLock{
+		fileName: name,
+	}
+}
+
+// Lock locks file.
+// If the file is already locked, the calling goroutine blocks until the file is unlocked.
+// If lock has been invoked without unlock, lock again will return an error.
+func (l *FileLock) Lock() error {
+	fd, err := os.Open(l.fileName)
+	if err != nil {
+		return err
+	}
+
+	l.mutex.Lock()
+	if l.fd != nil {
+		l.mutex.Unlock()
+		return fmt.Errorf("invoke Unlock() before lock %s again", l.fileName)
+	}
+	l.fd = fd
+	l.mutex.Unlock()
+
+	err = syscall.Flock(int(l.fd.Fd()), syscall.LOCK_EX)
+	if err != nil {
+		return fmt.Errorf("file %s lock failed error: %s", l.fileName, err)
+	}
+	return nil
+}
+
+// Unlock unlocks file.
+// If lock has not been invoked before unlock, unlock will return an error.
+func (l *FileLock) Unlock() error {
+	l.mutex.Lock()
+	if l.fd == nil {
+		l.mutex.Unlock()
+		return fmt.Errorf("file %s descriptor is nil", l.fileName)
+	}
+	fd := l.fd
+	l.fd = nil
+	l.mutex.Unlock()
+
+	defer fd.Close()
```

Comment:
What if fd.Close() has an error? Is it possible to return the error from it?

Reply:
no, I do not think so.

Reply:
Interesting. This **is** possible if it is using named return value.

```go
func (l *FileLock) Unlock() (err error) {
    ...
    defer func() {
        err = fd.Close()
    }()
    ...
}
```

But it is confusing anyway. 

1. What if `syscall.Flock` fails but `fd.Close()` succeeds?
1. What if `syscall.Flock` succeeds but `fd.Close()` fails?
1. What if both of them are failing?

Reply:
Yes, named return value can be changed in defer func indeed. Maybe we can merge the two errors in this func?  But it will look not so elegant.

Reply:
IMOP, I think we do not have to deal with all the returned errors, especially such as io.close(), file.close(), etc. And most of the similar cases in the golang source are handled in this way too.